src/lib/ppx_coda/versioned.ml

```diff
@@ -1,57 +1,174 @@
+    match num with
+    | Pexp_constant (Pconst_integer (n, None)) ->
+        if int_of_string n <= 0 then
+          Location.raise_errorf ~loc "Version must be a positive number" ;
+        true
+    | Pexp_constant _ ->
+        Location.raise_errorf ~loc "Version option must be a number"
+    | _ -> false
   in
-  match item with
-  | Pexp_ident id -> versioned_id id
-  | Pexp_apply ({pexp_desc= Pexp_ident id; _}, _) -> versioned_id id
+  match expr.pexp_desc with
+  | Pexp_apply
+      ( {pexp_desc= Pexp_ident id; _}
+      , [ ( Nolabel
+          , { pexp_desc=
+                Pexp_record
+                  ( [({txt= Lident "n"; _}, {pexp_desc= num; pexp_loc; _})]
+                  , None ); _ } ) ] ) ->
+      version_id id && is_version_number num pexp_loc
+  | Pexp_ident id ->
+      if version_id id then
+        Location.raise_errorf ~loc:expr.pexp_loc
+          "\"version\" requires a number option; example: \"{n = 1}\"" ;
+      false
   | _ -> false
 
-let payload_has_versioned payload =
+let payload_has_version payload =
   match payload with
   | PStr structure ->
       List.exists structure ~f:(fun str ->
           match str.pstr_desc with
           | Pstr_eval (expr, _) -> (
-            (* the "versioned" can appear as a singleton ident, or in a tuple *)
+            (* "version" can appear as an application or in a tuple *)
             match expr.pexp_desc with
-            | Pexp_ident _ -> is_versioned_ident expr.pexp_desc
-            | Pexp_apply ({pexp_desc; _}, _) -> is_versioned_ident pexp_desc
-            | Pexp_tuple items ->
-                List.exists items ~f:(fun item ->
-                    is_versioned_ident item.pexp_desc )
-            | _ -> false )
+            | Pexp_tuple exprs ->
+                List.exists exprs ~f:(fun expr -> is_version_application expr)
+            | _ -> is_version_application expr )
           | _ -> false )
   | _ -> false
 
-let attribute_has_deriving_versioned ((name, payload) : attribute) =
-  String.equal name.txt "deriving" && payload_has_versioned payload
+let attribute_has_deriving_version ((name, payload) : attribute) =
+  String.equal name.txt "deriving" && payload_has_version payload
+
+let attributes_have_deriving_version (attrs : attribute list) =
+  List.exists attrs ~f:attribute_has_deriving_version
+
+let type_has_deriving_version type_decl =
+  attributes_have_deriving_version type_decl.ptype_attributes
+
+let get_expr_version_string expr =
+  match expr.pexp_desc with
+  | Pexp_apply
+      ( {pexp_desc= Pexp_ident {txt= Lident "version"; _}; _}
+      , [ ( Nolabel
+          , { pexp_desc=
+                Pexp_record
+                  ( [ ( {txt= Lident "n"; _}
+                      , { pexp_desc=
+                            Pexp_constant
+                              (Pconst_integer (number_string, None)); _ } ) ]
+                  , None ); _ } ) ] ) ->
+      number_string
+  | _ -> failwith "Expected expression to be a version application"
```

Comment:
Nit: might be better to state exactly what we're expecting, e.g. `version expects an argument n for the version number, like version {n = 1}`

Reply:
if we extract the version from the module name, this becomes moot.